src/main/java/org/opensearch/knn/index/query/iterators/GroupedNestedDocIdSetIterator.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.query.iterators;
+
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.util.BitSet;
+
+import java.io.IOException;
+import java.util.ArrayList;
+
+/**
+ * A `DocIdSetIterator` that iterates over all nested document IDs belongs to the same parent document for a given
+ * set of nested document IDs.
+ *
+ * The {@link #docIds} should include only a single nested document ID per parent document. Otherwise, the nested documents
+ * of that parent document will be iterated multiple times.
+ *
+ * The order of document IDs in {@link #docIds} does not need to be sorted.
+
+ */
+public class GroupedNestedDocIdSetIterator extends DocIdSetIterator {
+    private final BitSet parentBitSet;
+    private final ArrayList<Integer> docIds;
```

Comment:
why not `List<Integer>`?

Reply:
Other form of list might cause performance degradation because we use `docIds.get(currentIndex)` a lot.